packages/orm/src/EntityManager.ts

```diff
@@ -482,30 +481,46 @@ export class EntityManager {
    * (i.e. including the initial `SELECT`s) in a transaction.
    */
   async flush(): Promise<void> {
-    if (this.currentFlushPromise) {
-      await this.currentFlushPromise;
-      await this.flushLoader.load(0);
-    } else {
-      await this.flushLoader.load(0);
+    if (this.isFlushing) {
+      throw new Error("Cannot flush while another flush is already in progress");
     }
-  }
 
-  /** The implementation of flush, but called by a DataLoader to de-dup calls made in a loop. */
-  private async doFlush(): Promise<void> {
-    // We defer doing this cascade logic until flush() so that delete() can remain synchronous.
-    await this.cascadeDeletesIntoUnloadedCollections();
-    recalcDerivedFields(this.entities);
-    const entityTodos = sortEntities(this.entities);
-    await addReactiveAsyncDerivedValues(entityTodos);
-    await recalcAsyncDerivedFields(this, entityTodos);
+    this._isFlushing = true;
+
+    // We need to split the thread so that Node's executionAsyncId changes and in turn
+    // Contexty give us a context that's distinct from whatever called flush
+    await new Promise((res) => setTimeout(res, 0));
+    const context = this.contexty.create();
+
+    const entitiesToFlush: Entity[] = [];
```

Comment:
Would we should make this a `Set` just for good measure.

Reply:
we could I suppose. The underlying `em.entities` is not a `Set`, though, so it would seem a little weird to make it one at this level instead.